essence-of-live-coding/src/LiveCoding/Migrate/NoMigration.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE Arrows #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveTraversable #-}
+
+{-|
+Module      : LiveCoding.Migrate.NoMigration
+Description : Mechanism to save state in a Cell without requiring a Data instance.
+
+If a data type is wrapped in 'NoMigration' then it can be used in a 'Cell' without requiring 
+it to have a 'Data' instance. The consequence is that if the type has changed
+in between a livereload, then the previous saved value will be discarded, and no migration attempt
+will happen.
+
+'LiveCoding' does not export 'delay' and 'changes' from this module. These functions should be
+used with a qualified import.
+-}
+module LiveCoding.Migrate.NoMigration where
+
+-- base
+import Data.Data
+    ( Data(dataTypeOf, gunfold, toConstr),
+      Typeable,
+      mkConstr,
+      mkDataType,
+      Constr,
+      DataType,
+      Fixity(Prefix) )
+
+-- essence-of-live-coding
+import LiveCoding.Cell ( Cell(Cell) )
+import LiveCoding.Cell.Monad ( hoistCellKleisli )
+import Control.Arrow ( Arrow(arr), returnA, (>>>) )
+import qualified LiveCoding.Cell.Feedback as Feedback
+import Control.Monad (guard)
+
+-- |Isomorphic to @'Maybe' a@ but has a different Data instance
+data NoMigration a = Initialized a | Uninitialized
+    deriving (Show, Eq, Functor, Foldable, Traversable)
+
+fromNoMigration :: a -> NoMigration a -> a
+fromNoMigration _ (Initialized a) = a
+fromNoMigration a Uninitialized = a
+
+dataTypeNoMigration :: DataType
+dataTypeNoMigration = mkDataType "NoMigration" [initializedConstr, uninitializedConstr]
+
+initializedConstr :: Constr
+initializedConstr = mkConstr dataTypeNoMigration "Initialized" [] Prefix
+
+uninitializedConstr :: Constr
+uninitializedConstr = mkConstr dataTypeNoMigration "Uninitialized" [] Prefix
+
+-- |This 'Data' instance is such that deserialization always returns an Uninitialized value.
+instance (Typeable a) => Data (NoMigration a) where
+  dataTypeOf _ = dataTypeNoMigration
+  toConstr (Initialized _) = initializedConstr
+  toConstr Uninitialized = uninitializedConstr
+  gunfold _cons nil _ = nil Uninitialized
+
+-- Should be identical to traverse'
```

Comment:
If it is, why do we need it?

Reply:
In fact, when do we ever need this function?

Reply:
The idea was to take a cell which needs the Data instance and turn it into a Cell which doesn't need the Data instance. I will re-check that it actually does that. 

At the moment traverse' hasn't been merged, so it is not available. The idea would be that once traverse' is merged this can be removed.

Reply:
Oh, that's https://github.com/turion/essence-of-live-coding/pull/95 which I've been neglecting. Thanks!

Reply:
Ok, let's merge #95 first then.

Reply:
> The idea was to take a cell which needs the Data instance and turn it into a Cell which doesn't need the Data instance. I will re-check that it actually does that.

Ah, yes, if I understand correctly you could check that by implementing some of the utility functions using `traverse'`?

Reply:
I have removed this function. Actually it is not useful. What we want to wrap in `NoMigration` is the state of the Cell and not the input or output value. Unfortunately a Cell with state s can't be converted to a Cell with state `NoMigration s` without requiring a `Data` instance for s, because to build the first Cell the `Data` instance is required.

Reply:
You could still write something like `cellNoMigration :: s -> (s -> a -> m (b, s)) -> Cell m a b`, which has the same type as the `Cell` constructor, but without the `Data` constraint.

Reply:
```Haskell
cellNoMigration :: (Typeable s, Functor m) => s -> (s -> a -> m (b, s)) -> Cell m a b
cellNoMigration state step = Cell { cellState = Uninitialized, ..} where
  cellStep Uninitialized a = second Initialized <$> step state a
  cellStep (Initialized s) a = second Initialized <$> step s a

cellNoMigration' :: Monad m => Cell m a b -> Cell m a b
cellNoMigration'(Cell state step) = cellNoMigration state step
cellNoMigration' cell = cell

changes2 :: (Monad m, Eq a, Typeable a) => Cell m a (Maybe a)
changes2 = cellNoMigration' Util.changes
```

```
• Could not deduce (Data a) arising from a use of ‘Util.changes’
  from the context: (Monad m, Eq a, Typeable a)
    bound by the type signature for:
               changes2 :: forall (m :: * -> *) a.
                           (Monad m, Eq a, Typeable a) =>
                           Cell m a (Maybe a)
    at /home/miguel/Development/Haskell/projects/SC/repos/essence-of-live-coding/essence-of-live-coding/src/LiveCoding/Migrate/NoMigration.hs:111:1-61
  Possible fix:
    add (Data a) to the context of
      the type signature for:
        changes2 :: forall (m :: * -> *) a.
                    (Monad m, Eq a, Typeable a) =>
                    Cell m a (Maybe a)
• In the first argument of ‘cellNoMigration'’, namely
    ‘Util.changes’
  In the expression: cellNoMigration' Util.changes
  In an equation for ‘changes2’:
      changes2 = cellNoMigration' Util.changes
```

If only there was a way to remove the `Data` constraint by some type trickery, or something like `coerce`... I didn't manage to find a way to do it. We want to use the exact same code (`cellStep` function) but without a constraint, which should be safe as we will never require using the associated functions of that constraint.

Reply:
Yeah, I think this won't work. The `Data` constraint is baked into half of the library, and you'll need it all the time when you call library functions.

If you want to share more code, we can make a new type:

```haskell
data Mealy m s a b = Mealy { s :: s, f :: a -> s -> m (b, s) }
cell :: Data s => Mealy m s a b -> Cell m a b
cell = _
cellNoMigration :: Mealy m s a b -> Cell m a b
cellNoMigration = _
```
Then all library functions that make sense in the "No-Migration-Context" can be written with this.